File: C++/9244 - Pinball.cpp
#include <algorithm>
#include <iostream>
#include <set>
#include <vector>

enum class Direction : int
{
    LEFT = 0,
    RIGHT = 1,
};

enum class EventType : int
{
    RIGHT_UP,
    RIGHT_DOWN,
    LEFT_UP,
    LEFT_DOWN,
};

struct Point
{
    Point() = default;
    Point(long long _x, long long _y) : x(_x), y(_y)
    {
        // Do nothing
    }

    long long x, y;
};

struct Interval
{
    Interval(int _idx, const Point& p1, const Point& p2) : idx(_idx)
    {
        if (p1.x <= p2.x)
        {
            l = p1;
            r = p2;
        }
        else
        {
            l = p2;
            r = p1;
        }

        if (l.y < r.y)
        {
            direction = Direction::LEFT;
        }
        else
        {
            direction = Direction::RIGHT;
        }
	}

    bool operator==(const Interval& interval) const
    {
        return idx == interval.idx;
    }

	bool operator<(const Interval& interval) const
    {
        if (idx == interval.idx)
        {
            return false;
        }

        if (IsBetween(interval.l))
		{
            return IsBelow(interval.l);
        }

        if (interval.IsBetween(l))
		{
            return !interval.IsBelow(l);
        }
        
        return idx < interval.idx;
    } 

    bool IsBetween(const Point& p) const
    {
        return (l.x <= p.x) && (p.x <= r.x);
    }

    bool IsBelow(const Point& p) const
    {
        return ((r.x - l.x) * (p.y - l.y) - (r.y - l.y) * (p.x - l.x)) < 0;
    }

	Point l, r;
	Direction direction;
	int idx;
};

struct Event
{
    Event(const Interval& interval, Direction side) : num(interval.idx)
    {
        if (side == Direction::LEFT)
        {
            x = interval.l.x;
        }
        else
        {
            x = interval.r.x;
        }
        
        if (interval.direction == Direction::LEFT && side == Direction::LEFT)
        {
            type = EventType::LEFT_DOWN;
        }
        else if (interval.direction == Direction::RIGHT && side == Direction::LEFT)
        {
            type = EventType::LEFT_UP;
        }
        else if (interval.direction == Direction::LEFT && side == Direction::RIGHT)
        {
            type = EventType::RIGHT_UP;
        }
        else if (interval.direction == Direction::RIGHT && side == Direction::RIGHT)
        {
            type = EventType::RIGHT_DOWN;
        }
    }

	bool operator<(const Event& e) const
    {
        if (x != e.x)
        {
            return x < e.x;
        }
        else
        {
            return type > e.type;
        }
	}

    EventType type;
    long long x;
    int num;
};

int main()
{
    int n;
    std::cin >> n;

    std::vector<Interval> input;
    std::vector<Event> events;
	
	for (int i = 0; i < n; ++i)
    {
		long long x1, y1, x2, y2;
        std::cin >> x1 >> y1 >> x2 >> y2;
		
		Interval interval{ i, Point{ x1, y1 }, Point{ x2, y2 } };
		input.emplace_back(interval);
		events.emplace_back(Event{ interval, Direction::LEFT });
		events.emplace_back(Event{ interval, Direction::RIGHT });
	}

	long long x0;
	std::cin >> x0;
	
	Interval interval{ n, Point{ x0 - 1, 2'000'001 }, Point{ x0, 2'000'000 } };
	input.emplace_back(interval);
    events.emplace_back(Event{ interval, Direction::LEFT });
    events.emplace_back(Event{ interval, Direction::RIGHT });
	
	std::sort(events.begin(), events.end());

    std::set<Interval> sweep;
	std::vector<int> succ(n + 1, -2);

	for (int i = 0; i < events.size(); ++i)
    {
		Event e = events[i];

		if (e.type == EventType::LEFT_UP)
        {
			sweep.insert(input[e.num]);
		}
		else if (e.type == EventType::RIGHT_UP)
        {
			sweep.erase(input[e.num]);
		}
		else if (e.type == EventType::LEFT_DOWN)
        {
			sweep.insert(input[e.num]);

			auto iter = sweep.find(input[e.num]);
            ++iter;

			if (iter == sweep.end())
            {
                succ[e.num] = -1;
            }
            else
            {
                succ[e.num] = iter->idx;
            }
		}
		else if (e.type == EventType::RIGHT_DOWN)
        {
			auto iter = sweep.find(input[e.num]);
            ++iter;

			if (iter == sweep.end())
            {
                succ[e.num] = -1;
            }
            else
            {
                succ[e.num] = iter->idx;
            }

			sweep.erase(input[e.num]);
		}
	}

	int query = n;
	long long xPos = 0;

	while (succ[query] >= 0)
    {
        query = succ[query];
    }

	if (input[query].direction == Direction::LEFT)
    {
        xPos = input[query].l.x;
    }
    else
    {
        xPos = input[query].r.x;
    }

	std::cout << xPos << '\n';

	return 0;
}

File: Rust/9244 - Pinball.rs
use io::Write;
use std::{collections::BTreeSet, io, str};

pub struct UnsafeScanner<R> {
    reader: R,
    buf_str: Vec<u8>,
    buf_iter: str::SplitAsciiWhitespace<'static>,
}

impl<R: io::BufRead> UnsafeScanner<R> {
    pub fn new(reader: R) -> Self {
        Self {
            reader,
            buf_str: vec![],
            buf_iter: "".split_ascii_whitespace(),
        }
    }

    pub fn token<T: str::FromStr>(&mut self) -> T {
        loop {
            if let Some(token) = self.buf_iter.next() {
                return token.parse().ok().expect("Failed parse");
            }
            self.buf_str.clear();
            self.reader
                .read_until(b'\n', &mut self.buf_str)
                .expect("Failed read");
            self.buf_iter = unsafe {
                let slice = str::from_utf8_unchecked(&self.buf_str);
                std::mem::transmute(slice.split_ascii_whitespace())
            }
        }
    }
}

#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
struct Point {
    x: i64,
    y: i64,
}

impl Point {
    fn new(x: i64, y: i64) -> Self {
        Self { x, y }
    }
}

#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
enum Direction {
    Left,
    Right,
}

#[derive(Clone, Copy, PartialOrd)]
struct Interval {
    left: Point,
    right: Point,
    direction: Direction,
    idx: usize,
}

impl Interval {
    fn new(idx: usize, p1: Point, p2: Point) -> Self {
        let left = if p1.x <= p2.x { p1 } else { p2 };
        let right = if p1.x <= p2.x { p2 } else { p1 };
        let direction = if left.y < right.y {
            Direction::Left
        } else {
            Direction::Right
        };

        Self {
            left,
            right,
            direction,
            idx,
        }
    }

    fn is_between(&self, p: &Point) -> bool {
        self.left.x <= p.x && p.x <= self.right.x
    }

    fn is_below(&self, p: &Point) -> bool {
        ((self.right.x - self.left.x) * (p.y - self.left.y)
            - (self.right.y - self.left.y) * (p.x - self.left.x))
            < 0
    }
}

impl PartialEq for Interval {
    fn eq(&self, other: &Self) -> bool {
        self.idx == other.idx
    }
}

impl Eq for Interval {}

impl Ord for Interval {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        if self.idx == other.idx {
            return std::cmp::Ordering::Equal;
        }

        if self.is_between(&other.left) {
            if self.is_below(&other.left) {
                return std::cmp::Ordering::Less;
            } else {
                return std::cmp::Ordering::Greater;
            }
        }

        if other.is_between(&self.left) {
            if other.is_below(&self.left) {
                return std::cmp::Ordering::Greater;
            } else {
                return std::cmp::Ordering::Less;
            }
        }

        if self.idx < other.idx {
            std::cmp::Ordering::Less
        } else {
            std::cmp::Ordering::Greater
        }
    }
}

#[derive(Clone, Copy, PartialOrd, Ord, PartialEq, Eq)]
enum EventType {
    RightUp,
    RightDown,
    LeftUp,
    LeftDown,
}

struct Event {
    event_type: EventType,
    x: i64,
    num: i64,
}

impl Event {
    fn new(interval: Interval, side: Direction) -> Self {
        Self {
            event_type: match side {
                Direction::Left => match interval.direction {
                    Direction::Left => EventType::LeftDown,
                    Direction::Right => EventType::LeftUp,
                },
                Direction::Right => match interval.direction {
                    Direction::Left => EventType::RightUp,
                    Direction::Right => EventType::RightDown,
                },
            },
            x: match side {
                Direction::Left => interval.left.x,
                Direction::Right => interval.right.x,
            },
            num: interval.idx as i64,
        }
    }
}

fn main() {
    let (stdin, stdout) = (io::stdin(), io::stdout());
    let mut scan = UnsafeScanner::new(stdin.lock());
    let mut out = io::BufWriter::new(stdout.lock());

    let n = scan.token::<usize>();
    let mut input = Vec::new();
    let mut events = Vec::new();

    for i in 0..n {
        let (x1, y1, x2, y2) = (
            scan.token::<i64>(),
            scan.token::<i64>(),
            scan.token::<i64>(),
            scan.token::<i64>(),
        );
        let interval = Interval::new(i, Point::new(x1, y1), Point::new(x2, y2));

        input.push(interval);
        events.push(Event::new(interval, Direction::Left));
        events.push(Event::new(interval, Direction::Right));
    }

    let x0 = scan.token::<i64>();
    let interval = Interval::new(n, Point::new(x0 - 1, 2_000_001), Point::new(x0, 2_000_000));

    input.push(interval);
    events.push(Event::new(interval, Direction::Left));
    events.push(Event::new(interval, Direction::Right));

    events.sort_by(|a, b| {
        if a.x != b.x {
            a.x.cmp(&b.x)
        } else {
            b.event_type.cmp(&a.event_type)
        }
    });

    let mut sweep = BTreeSet::new();
    let mut succ: Vec<i64> = vec![-2; n + 1];

    for event in events.iter() {
        println!("{}", event.num);
        let event_num = event.num as usize;

        match event.event_type {
            EventType::LeftUp => {
                sweep.insert(input[event_num]);
            }
            EventType::RightUp => {
                sweep.remove(&input[event_num]);
            }
            EventType::LeftDown => {
                sweep.insert(input[event_num]);

                if let Some(val) = sweep.iter().last() {
                    if val == &input[event_num] {
                        succ[event_num] = -1;
                    } else {
                        succ[event_num] = val.idx as i64;
                    }
                }
            }
            EventType::RightDown => {
                if let Some(val) = sweep.iter().last() {
                    if val == &input[event_num] {
                        succ[event_num] = -1;
                    } else {
                        succ[event_num] = val.idx as i64;
                    }
                }

                sweep.remove(&input[event_num]);
            }
        }
    }

    let mut query = n;

    while succ[query] >= 0 {
        query = succ[query] as usize;
    }

    let x_pos = if input[query].direction == Direction::Left {
        input[query].left.x
    } else {
        input[query].right.x
    };

    writeln!(out, "{}", x_pos).unwrap();
}
ㅌ